### ISSUE TITLE
AWX EE performance

### ISSUE THREAD

**Issue by adpavlov** [ansible/awx#12088]
##### ISSUE TYPE
 - Feature Idea

##### SUMMARY

Hi everyone,

Finally trying to migrate my docker based AWX 15 to kubernetes based AWX 20.
I have to use custom EE to have ansible 2.9.
In EE ideology each playbook execution means separate pod/container creation, which actually takes a lot of time to bring infrastructure up before actual execution starts.
So I have few questions that I could not google an answer:

1. Can I make custom EE pod always ready and running like it's done for control plane EE?
2. Can I make e.g. 4 containers always up and ready to start execution immediately?

Thanks in advance!

**Comment by kdelee** [comment]
@adpavlov that is not currently a feature. I would guess you might need some changes to the receptor kubernetes work type if it were to be one. https://github.com/ansible/receptor/blob/devel/docs/source/k8s.rst

In OpenShift there is a registry local to the cluster, so once the image is available there, the start up time is not much of an issue.
You may consider looking at setting up an internal registry on k8s.
https://blog.container-solutions.com/installing-a-registry-on-kubernetes-quickstart

I'm going to close this issue as it is not a specific feature request for awx. This type of discussion is better conducted on the mailing list. 
https://groups.google.com/g/awx-project

**Comment by adpavlov** [comment]
@kdelee thanks for looking into it. Probably I didn't explained it wright.
I have custom AWX-EE image stored locally as my environment is isolated from the internet, but still it takes quite a lot of time to create POD, container and then start job execution.
I want to achieve constantly running and ready to execute containers like it's illustrated in [Docker Compose for Development](https://github.com/ansible/awx/blob/devel/tools/docker-compose/README.md)
![image](https://user-images.githubusercontent.com/51116816/164726066-344651ee-f44a-4ce5-b756-e11da5e47618.jpeg) or like it was in docker architecture for AWX 15, where awx_task container was always up and handled same playbooks quite fast.

Now I see quite strange behavior that awx_task container first overloading entire physical server with git processes:

```
   PID USER      PR  NI    VIRT    RES    SHR S  %CPU %MEM     TIME+ COMMAND                                                                       
   529 1000      20   0  341196  93764    392 S  37.5  1.6   0:37.20 git                                                                           
   375 1000      20   0  341196  82636    392 R  35.2  1.4   0:36.71 git                                                                           
127853 root      20   0 1392532 441908  37736 S  34.2  7.5  28:31.24 k3s-server                                                                    
   385 1000      20   0  374024 123256    392 R  28.9  2.1   0:38.26 git                                                                           
   515 1000      20   0  341196 134380    396 R  28.6  2.3   0:36.47 git                                                                           
   499 1000      20   0  382604 115740    392 S  27.9  2.0   0:34.10 git                                                                           
   541 1000      20   0  349776 171496    396 R  26.2  2.9   0:37.08 git                                                                           
   522 1000      20   0  396300 272812    392 R  24.9  4.6   0:38.43 git                                                                           
   562 1000      20   0  341112 153064  17172 R  24.3  2.6   0:38.87 git                                                                           
   646 1000      20   0  341196 149272    396 R  23.9  2.5   0:39.69 git                                                                           
  2206 1000      20   0   69116  50276   7496 S  19.9  0.8   0:00.60 awx-man

[root@rhel7 ~]# ps -aef|grep git
1000        372  12139  0 09:40 ?        00:00:00 git clone --branch=awx_internal/94cc22aa-a47a-4ef9-9d11-7f8651e8706e --depth=1 --single-branch -v file:///var/lib/awx/projects/_8__rhel7 /tmp/awx_35_fzlid_0l/project
1000        373    372  1 09:40 ?        00:00:02 git-upload-pack /var/lib/awx/projects/_8__rhel7
1000        375    373 22 09:40 ?        00:00:40 /usr/libexec/git-core/git --shallow-file  pack-objects --revs --thin --stdout --shallow --delta-base-offset --include-tag
1000        377    372  3 09:40 ?        00:00:06 /usr/libexec/git-core/git --shallow-file /tmp/awx_35_fzlid_0l/project/.git/shallow.lock index-pack --stdin --fix-thin --keep=fetch-pack 521 on awx-7855b8944f-rfh4b
1000        382  12139  0 09:40 ?        00:00:01 git clone --branch=awx_internal/72578585-5d5c-49d2-a23c-ae260c6021f4 --depth=1 --single-branch -v file:///var/lib/awx/projects/_8__rhel7 /tmp/awx_36_77ps9ii_/project
1000        497  12139  0 09:40 ?        00:00:00 git clone --branch=awx_internal/e3e6e027-2255-4a75-b3f3-1640525be43a --depth=1 --single-branch -v file:///var/lib/awx/projects/_8__rhel7 /tmp/awx_40_71jxtlr5/project
1000        498    497  1 09:40 ?        00:00:02 git-upload-pack /var/lib/awx/projects/_8__rhel7
1000        499    498 21 09:40 ?        00:00:37 /usr/libexec/git-core/git --shallow-file  pack-objects --revs --thin --stdout --shallow --delta-base-offset --include-tag
1000        501    497  3 09:40 ?        00:00:06 /usr/libexec/git-core/git --shallow-file /tmp/awx_40_71jxtlr5/project/.git/shallow.lock index-pack --stdin --fix-thin --keep=fetch-pack 566 on awx-7855b8944f-rfh4b
1000        512  12139  0 09:40 ?        00:00:01 git clone --branch=awx_internal/dc534c4a-0eb2-491f-8fa7-039ee274d206 --depth=1 --single-branch -v file:///var/lib/awx/projects/_8__rhel7 /tmp/awx_43_phcjkltf/project
1000        520  12139  0 09:40 ?        00:00:00 git clone --branch=awx_internal/132476e9-0a9a-4097-aa2c-23a3f15f10e2 --depth=1 --single-branch -v file:///var/lib/awx/projects/_8__rhel7 /tmp/awx_41_iq4z3nm9/project
1000        521    520  1 09:40 ?        00:00:02 git-upload-pack /var/lib/awx/projects/_8__rhel7
1000        522    521 24 09:40 ?        00:00:43 /usr/libexec/git-core/git --shallow-file  pack-objects --revs --thin --stdout --shallow --delta-base-offset --include-tag
1000        524    520  3 09:40 ?        00:00:06 /usr/libexec/git-core/git --shallow-file /tmp/awx_41_iq4z3nm9/project/.git/shallow.lock index-pack --stdin --fix-thin --keep=fetch-pack 585 on awx-7855b8944f-rfh4b
1000        527  12139  0 09:40 ?        00:00:00 git clone --branch=awx_internal/8f778a6c-466a-4dae-aa7a-e1f1a2594567 --depth=1 --single-branch -v file:///var/lib/awx/projects/_8__rhel7 /tmp/awx_42_1f3_pj0e/project
1000        539  12139  0 09:40 ?        00:00:00 git clone --branch=awx_internal/e972331f-9076-4171-967a-d9fe8f4c7fb9 --depth=1 --single-branch -v file:///var/lib/awx/projects/_8__rhel7 /tmp/awx_44_di47hs73/project
1000        540    539  1 09:40 ?        00:00:01 git-upload-pack /var/lib/awx/projects/_8__rhel7
1000        541    540 23 09:40 ?        00:00:41 /usr/libexec/git-core/git --shallow-file  pack-objects --revs --thin --stdout --shallow --delta-base-offset --include-tag
1000        543    539  3 09:40 ?        00:00:05 /usr/libexec/git-core/git --shallow-file /tmp/awx_44_di47hs73/project/.git/shallow.lock index-pack --stdin --fix-thin --keep=fetch-pack 601 on awx-7855b8944f-rfh4b
1000        560  12139  0 09:40 ?        00:00:00 git clone --branch=awx_internal/eff54209-358f-457b-8a89-1768940a482b --depth=1 --single-branch -v file:///var/lib/awx/projects/_8__rhel7 /tmp/awx_46_va_c3dfq/project
1000        561    560  1 09:40 ?        00:00:02 git-upload-pack /var/lib/awx/projects/_8__rhel7
1000        564    560  3 09:40 ?        00:00:06 /usr/libexec/git-core/git --shallow-file /tmp/awx_46_va_c3dfq/project/.git/shallow.lock index-pack --stdin --fix-thin --keep=fetch-pack 614 on awx-7855b8944f-rfh4b
1000        632  12139  0 09:40 ?        00:00:00 git clone --branch=awx_internal/6fa7a914-11e8-493d-bd5e-8a2336ca1979 --depth=1 --single-branch -v file:///var/lib/awx/projects/_8__rhel7 /tmp/awx_47_eg9yg3ou/project
1000        645    632  1 09:40 ?        00:00:02 git-upload-pack /var/lib/awx/projects/_8__rhel7
1000        646    645 24 09:40 ?        00:00:43 /usr/libexec/git-core/git --shallow-file  pack-objects --revs --thin --stdout --shallow --delta-base-offset --include-tag
1000        648    632  3 09:40 ?        00:00:06 /usr/libexec/git-core/git --shallow-file /tmp/awx_47_eg9yg3ou/project/.git/shallow.lock index-pack --stdin --fix-thin --keep=fetch-pack 669 on awx-7855b8944f-rfh4b
```

Then some jobs are just failing without any output (I need to dig here more, but I saw some timeouts from redis container).
Others taking like double time for execution compared to dockerized AWX 15.

For example I have 16-core Xeon Server with 32 Gb of RAM. I executed like 10 playbooks in parallel and it took 15 minutes to just skip tasks as no hosts matched. 

![Снимок экрана 2022-04-20 в 01 09 46](https://user-images.githubusercontent.com/51116816/164724658-a886c7c6-ccf1-4c2c-a74e-963707ef0f37.jpg)

**Comment by kdelee** [comment]
The always-on-receptor nodes you are describing in the development environment are simulating a type of execution instance which is not currently a part of the mainline awx deployment on k8s, which only uses ContainerGroups for job execution. Container groups use that k8s work type in receptor https://github.com/ansible/receptor/blob/devel/docs/source/k8s.rst

re: the git processes
you might check your project's "update on launch" option
